src/main/kotlin/org/rust/ide/template/postfix/utils.kt

```diff
@@ -28,48 +28,32 @@ internal object RsPostfixTemplatePsiInfo : PostfixTemplatePsiInfo() {
         RsPsiFactory(context.project).createExpression("$prefix${context.text}$suffix")
 }
 
-abstract class RsExprParentsSelectorBase(val pred: (RsExpr) -> Boolean) : PostfixTemplateExpressionSelector {
-    override fun getRenderer(): Function<PsiElement, String> = Function { it.text }
-
+abstract class RsExprParentsSelectorBase(pred: (RsExpr) -> Boolean) : PostfixTemplateExpressionSelectorBase(Condition { it is RsExpr && pred(it) }) {
     final override fun getExpressions(context: PsiElement, document: Document, offset: Int): List<PsiElement> {
-        val expressions = getExpressionsInternal(context, document, offset)
+        val expressions = super.getExpressions(context, document, offset)
         // `PostfixTemplateWithExpressionSelector#expand` selects only one item from this list in unit tests.
         // But in different platform versions different items are selected (of course, it's very convenient).
         // So let's return the latest item to commit tests behavior with all platform versions
         return if (isUnitTestMode) listOfNotNull(expressions.lastOrNull()) else expressions
     }
-
-    protected abstract fun getExpressionsInternal(context: PsiElement, document: Document, offset: Int): List<PsiElement>
 }
 
 class RsTopMostInScopeSelector(pred: ((RsExpr) -> Boolean) = { true }) : RsExprParentsSelectorBase(pred) {
-    override fun getExpressionsInternal(context: PsiElement, document: Document, offset: Int): List<PsiElement> =
+    override fun getNonFilteredExpressions(context: PsiElement, document: Document, offset: Int): List<PsiElement> =
         context
             .ancestors
             .takeWhile { it !is RsBlock && it.endOffset == context.endOffset }
```

Comment:
I'm not sure if `it.endOffset == context.endOffset` condition is still needed. All tests in `org.rust.ide.template.postfix` package pass without it.

Reply:
Oh, it's interesting

Currently, some of our postfix templates behave in the following way:
![2021-03-22 18 25 55](https://user-images.githubusercontent.com/2539310/112014558-1154c980-8b3c-11eb-94af-15efc5043cc9.gif)

And it's rather weird in my opinion that the plugin tries to select all parent expressions in a block even if we are inside this expression. The same postfix templates work differently in other languages.

But after your changes, they become to works in a similar way as in other languages. At the same time, `RsTopMostInScopeSelector` and `RsAllParentsSelector` provide the same behaviour now because `PostfixTemplateExpressionSelectorBase` always use `element.getTextRange().getEndOffset() == offset` condition so I suppose we can merge them into single `RsExprParentsSelector`

Reply:
@Undin fixed.